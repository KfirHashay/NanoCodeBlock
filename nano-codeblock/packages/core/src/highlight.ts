import Prism from 'prismjs';

// ensure Prism is attached to global scope for language modules
if (typeof globalThis !== 'undefined') {
  (globalThis as unknown as { Prism: typeof Prism }).Prism = Prism;
}

// load only required languages
import 'prismjs/components/prism-javascript';
import 'prismjs/components/prism-typescript';
import 'prismjs/components/prism-bash';
import 'prismjs/components/prism-json';

export type Token = {
  content: string;
  type: string;
};

function normalizeTokens(tokens: (string | Prism.Token)[]): Token[] {
  return tokens.map((token) => {
    if (typeof token === 'string') {
      return { content: token, type: 'plain' };
    }
    return { content: String(token.content), type: token.type };
  });
}

function splitTokens(tokens: (string | Prism.Token)[]): Token[][] {
  const lines: Token[][] = [[]];

  for (const token of tokens) {
    if (typeof token === 'string') {
      const parts = token.split(/\n/);
      lines[lines.length - 1].push({ content: parts[0], type: 'plain' });
      for (let i = 1; i < parts.length; i++) {
        lines.push([]);
        if (parts[i]) {
          lines[lines.length - 1].push({ content: parts[i], type: 'plain' });
        }
      }
    } else {
      const value = token.content;
      const content = typeof value === 'string' ? value : String(value);
      lines[lines.length - 1].push({ content, type: token.type });
    }
  }

  return lines.map(normalizeTokens);
}

export function highlight(code: string, lang: string): Token[][] {
  const grammar = Prism.languages[lang] ?? Prism.languages.javascript;
  const rawTokens = Prism.tokenize(code, grammar);
  return splitTokens(rawTokens);
}
