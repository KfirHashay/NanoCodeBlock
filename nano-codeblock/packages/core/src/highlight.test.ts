import { describe, it, expect } from 'vitest';
import { highlight } from './highlight';

const snippets: Record<string, string> = {
  javascript: 'const x = 1;\nconsole.log(x);',
  typescript: 'const x: number = 1;\nconsole.log(x);',
  bash: 'echo "hello"',
  json: '{"a":1}',
};

describe('highlight', () => {
  for (const [lang, code] of Object.entries(snippets)) {
    it(`tokenizes ${lang}`, () => {
      expect(highlight(code, lang)).toMatchSnapshot();
    });
  }
});
