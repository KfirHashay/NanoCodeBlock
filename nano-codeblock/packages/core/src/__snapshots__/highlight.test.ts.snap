// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`highlight > tokenizes bash 1`] = `
[
  [
    {
      "content": "echo",
      "type": "builtin",
    },
    {
      "content": " ",
      "type": "plain",
    },
    {
      "content": ""hello"",
      "type": "string",
    },
  ],
]
`;

exports[`highlight > tokenizes javascript 1`] = `
[
  [
    {
      "content": "const",
      "type": "keyword",
    },
    {
      "content": " x ",
      "type": "plain",
    },
    {
      "content": "=",
      "type": "operator",
    },
    {
      "content": " ",
      "type": "plain",
    },
    {
      "content": "1",
      "type": "number",
    },
    {
      "content": ";",
      "type": "punctuation",
    },
    {
      "content": "",
      "type": "plain",
    },
  ],
  [
    {
      "content": "console",
      "type": "plain",
    },
    {
      "content": ".",
      "type": "punctuation",
    },
    {
      "content": "log",
      "type": "function",
    },
    {
      "content": "(",
      "type": "punctuation",
    },
    {
      "content": "x",
      "type": "plain",
    },
    {
      "content": ")",
      "type": "punctuation",
    },
    {
      "content": ";",
      "type": "punctuation",
    },
  ],
]
`;

exports[`highlight > tokenizes json 1`] = `
[
  [
    {
      "content": "{",
      "type": "punctuation",
    },
    {
      "content": ""a"",
      "type": "property",
    },
    {
      "content": ":",
      "type": "operator",
    },
    {
      "content": "1",
      "type": "number",
    },
    {
      "content": "}",
      "type": "punctuation",
    },
  ],
]
`;

exports[`highlight > tokenizes typescript 1`] = `
[
  [
    {
      "content": "const",
      "type": "keyword",
    },
    {
      "content": " x",
      "type": "plain",
    },
    {
      "content": ":",
      "type": "operator",
    },
    {
      "content": " ",
      "type": "plain",
    },
    {
      "content": "number",
      "type": "builtin",
    },
    {
      "content": " ",
      "type": "plain",
    },
    {
      "content": "=",
      "type": "operator",
    },
    {
      "content": " ",
      "type": "plain",
    },
    {
      "content": "1",
      "type": "number",
    },
    {
      "content": ";",
      "type": "punctuation",
    },
    {
      "content": "",
      "type": "plain",
    },
  ],
  [
    {
      "content": "console",
      "type": "builtin",
    },
    {
      "content": ".",
      "type": "punctuation",
    },
    {
      "content": "log",
      "type": "function",
    },
    {
      "content": "(",
      "type": "punctuation",
    },
    {
      "content": "x",
      "type": "plain",
    },
    {
      "content": ")",
      "type": "punctuation",
    },
    {
      "content": ";",
      "type": "punctuation",
    },
  ],
]
`;
